# Data Wrangling with Lytics Profile Data - Tools and Techniques

The goal of this notebook is to present some tools and techniques that can be used to wrangle Industry Dive data. 

## What is Data Wrangling again?
>Data wrangling, sometimes referred to as data munging, is the process of transforming and mapping data from one "raw" data form into another format with the intent of making it more appropriate and valuable for a variety of downstream purposes such as analytics.  Some transformation techniques include: parsing, joining, standardizing, augmenting, cleansing, and consolidating. 

[per wikipedia](https://en.wikipedia.org/wiki/Data_wrangling)

## Bad Data in, Bad Data out

![bad data in bad data out](https://cdn-images-1.medium.com/max/1200/0*YCghEemt6BtW9OZV.png "Bad Data in Bad Data out")

Many websites contain forms in order to collect information from users for various reasons.  In our case, we have signup forms for dives that asks for information about our users like so:

![signup form](../data/img/signup_form.png "signup form")

As you can see, there are fields that are restricted to pre-defined values (e.g., Job Function), and free-form fields (e.g., Company Name) where a user can type most anything they like.  Whenever users are exposed to free-form fields, there is a possibility of bad/messy/non-standardized data making into your system.

For example, here are some variants of "IKEA" that are present for user profiles that we have:

* IKEA
* IKEA AG
* IKEA Belgium
* IKEA Canada
* IKEA Danville
* IKEA Food
* IKEA Home Furnishings
* IKEA Portugal
* IKEA USA
* IKEA US EAST, LLC 215
* IKEA US

Without some wrangling, you would not be able to aggregate these folks properly into a single group based on company.

## Lytics Profile Data
Now, let's take a look at some Lytics profile data, which consists of all information we have about users who interact with our content.  Within this data, there are key demographic fields that can help us understand who our users are, such as:
* first and last name
* job title
* email domain
* company name
* address

The data file we are going to look at is an export of the "All" audience segment in Lytics.
https://activate.getlytics.com/audiences/4cc5d612f46fb86e5cfd0c995250e60c/summary?aid=2751

![All Audience segment in Lytics](../data/img/lytics_all_audience_segment.png "All Audience segment in Lytics")

Let's start looking at this data to see how we can clean it up in order to help us create more accurate statistics about our users.

In [11]:
import pandas as pd
import numpy as np

dtypes = {'company': 'str', 'company_name': 'str', 'domain': 'object', 'emaildomain': 'object', 'emaildomains': 'object',
         'st_profile_id': 'object', 'user_id': np.float64, 'lytics_segment': 'object'}
df = pd.read_csv('../data/files/lytics_profile_data_export.csv', sep=',', error_bad_lines=False, index_col=False, dtype=dtypes)

# list columns in dataset
print(list(df))

# number of rows
print('# of rows left: %s' % df.shape[0])
# print(df[df['st_profile_id'].str.contains("5a2ba1f6ff530ac11a8b4868", na=False)])

['company', 'company_name', 'domain', 'emaildomain', 'emaildomains', 'st_profile_id', 'user_id', 'lytics_segment']
# of rows left: 782425


There are multiple fields in the data we can choose to cleanup, but first let's look at the "company_name" field.  One of the first things we should do is get rid of rows with company name values we don't care about.

In [12]:
# remove null company name values
df = df.dropna(subset=['company_name'])

# number of rows
print('# of rows left: %s' % df.shape[0])

# of rows left: 458289


In [13]:
# find values that are any combination of special characters
special_char_values = df['company_name'].str.contains("^[!@#$%^&*(),.?]*$", na=False)
print(df[special_char_values].company_name.unique())

# number of rows
print('# of special character value rows: %s' % df[special_char_values].shape[0])
df = df[~special_char_values]

print('# of rows left: %s' % df.shape[0])
# print(df[df['st_profile_id'].str.contains("5a2ba1f6ff530ac11a8b4868", na=False)])

['..' '.' '...' '*' '********' '......' ',' '.....' '***' '????????' '?'
 '**' '.......' ',,' '@@']
# of special character value rows: 103
# of rows left: 458186


In [14]:
# find values that are only numbers
number_values = df['company_name'].str.contains("^[0-9]*$", na=False)
print(df[number_values].company_name.unique())

# number of rows
print('# of number value rows: %s' % df[number_values].shape[0])
df = df[~number_values]

print('# of rows left: %s' % df.shape[0])

['1948' '1989' '1954' '451' '1957' '1979' '252' '1953' '1967' '8020'
 '1960' '5' '104' '1999' '123' '1974' '1988' '1977' '1000' '900' '1956'
 '605' '8760' '1984' '1959' '1998' '1972' '1992' '1997' '1991' '111'
 '1990' '1987' '1970' '1969' '1965' '1968' '1995' '1993' '1975' '1963'
 '231112027' '53' '1976' '1985' '1949' '149' '0' '1971' '1986' '346'
 '47723' '1947' '94122202312' '1' '1958' '1973' '43' '1935' '1961' '1994'
 '1946' '325024080134' '1996' '1982' '15' '34' '1952' '271' '1980' '1966'
 '1936' '47' '1978' '1964' '1928' '50' '2714' '1955' '1690' '1942' '13'
 '05358359981' '9172077326' '12' '151' '1951' '2000' '400000000000' '2'
 '1905' '2020' '1940' '1983' '2008' '198' '2013' '1962' '411' '2015' '295'
 '1950' '940005848995' '11455' '83255804' '2166833' '1001' '6' '91957'
 '14' '887000000000' '666' '59' '963' '32000' '555' '404' '0789243438'
 '438' '68' '1945' '525' '825' '2009' '1981' '8001504151' '136' '359'
 '365' '308' '940003979987' '6164381822' '1107' '0673282495' '2040' '74

In [15]:
# random additional values that I found when I was looking at the data in Excel
weird_vals = ['#NAME?', '{Re}', '< self >']
weird_values = df['company_name'].isin(weird_vals)
df = df[~weird_values]

# left over rows in dataframe
print('# of rows left: %s' % df.shape[0])

# of rows left: 457462


Now that we have cleaned all the bad company name values from our dataset, let's work on standardizing the names to help with comparison.

In [16]:
# change the values to all lower case
df['stndrdzed_company_name'] = df['company_name'].str.lower()
# remove all punctuation
df["stndrdzed_company_name"] = df['stndrdzed_company_name'].str.replace('[^\w\s]','')

# remove rows with "none" as value
none_rows = df['stndrdzed_company_name'].str.contains('none', na=False)
df = df[~none_rows]

# remove rows with "" as value
empty_string_rows = df['stndrdzed_company_name'].values == ''
df = df[~empty_string_rows]
print('# of rows left: %s' % df.shape[0])

# of rows left: 456521


Let's take a look at our dataset to see what we are working with:

In [17]:
grouped = df.groupby('stndrdzed_company_name')

grouped = grouped.size().reset_index(name='counts')
grouped.sort_values(by=['counts'], ascending=False)

stndrdzed_company_name  counts
449                                    20160506deleteme    1364
214673                                             self     692
115522                                              ibm     545
263824                                          walmart     523
3715                                          accenture     512
147255                                            macys     444
230752                                          student     414
72555                                       duke energy     412
163621                                               mr     384
214692                                    self employed     379
235724                                           target     378
173896                                         novartis     350
203159                                          retired     338
264512                                 waste management     335
196171                                              pwc     325
186641                                           pfizer     324
65683                                          deloitte     317
100383                                           google     317
56303                                        consultant     316
12326                                            amazon     313
271713                                      xcel energy     304
202363                                republic services     303
224607                       southern california edison     292
13549                                             amgen     290
166903                                    national grid     288
179351                                           oracle     287
210887                                           sanofi     285
12831                           american electric power     285
185693                                          pepsico     276
84585                                                ey     275
...                                                 ...     ...
99852                               golden rays of hope       1
99851                                golden pony casino       1
99836                        golden nugget lake charles       1
99821                                      golden homes       1
99822                              golden hour farm llc       1
99823                      golden isles medical billing       1
99824                                      golden krust       1
99825                               golden krust bakery       1
99826                 golden krust caribbean bakery inc       1
99829                     golden living village gardens       1
99830                     golden media service provider       1
99832                       golden nugget atlantic city       1
99833                              golden nugget biloxi       1
99834                        golden nugget casino hotel       1
99835                         golden nugget flea market       1
99837                              golden nugget online       1
99850                           golden polytech pvt ltd       1
99838                       golden olives academy abuja       1
99839                           golden one credit union       1
99840                              golden opportunities       1
99841                                       golden pack       1
99842                                     golden palace       1
99843                              golden palm advisors       1
99844                                   golden pass lng       1
99845   golden pasta flour mills of nigeria plc nigeria       1
99846                         golden peacock consulting       1
99847                          golden peanut  tree nuts       1
99848                       golden peanut and tree nuts       1
99849                                    golden platter       1
275265                                          현대엔지니어링       1

[275266 rows x 2 columns]

One thing to note from looking at this is that there are company names that contain values other than English.  For instance, "현대엔지니어링" is Korean.  This is one thing you could work on eliminating as well if you wanted to focus on English values.  I tried to use a library called "langdetect" for this, but it did not do a good job of picking up the obvious cases.

Once we have wrangled the data bit, we can now try to enhance our dataset with an external dataset.  One of the datasets we bought rights to recently, DiscoverOrg, has different information about companies that could be useful for analysis.  The common field these two datasets have is the company name.  So we can try to load this dataset, clean it up a bit, then compare it to our original cleaned dataset in order to try and match on company name and enhance our existing dataset.

In [18]:
dtypes= {'Company ID': np.int64, 'Company Name': 'str', 'Company Website': 'object', 'Company HQ Phone': 'object',
        'Company Email Domain': 'object', 'Company Description': 'object', 'Company Primary Industry': 'object',
        'Company Revenue': np.float64, 'Company IT Budget (Mil)': 'object', 'Number of Employees': np.int64,
        'Company IT Employees': np.float64, 'Company Fortune Rank': np.float64, 'Company Ownership': 'object', 'Company Profile URL': 'object',
        'Company Business Model (B2B/B2C/B2G)': 'object', 'Hospital Beds': 'object', 'HQ Address 1': 'object', 'HQ Address 2': 'object',
        'HQ City': 'object', 'HQ State': 'object', 'HQ Postal Code': 'object', 'HQ County': 'object', 'HQ Country': 'object'
        }
df2 = pd.read_csv('../data/files/DiscoverOrg_Company_223030_20180731141156.csv', encoding='latin-1', sep=',', error_bad_lines=False, index_col=False, dtype=dtypes)

# change the values to all lower case
df2['stndrdzed_company_name'] = df2['Company Name'].astype(str).str.lower()
# remove all punctuation
df2["stndrdzed_company_name"] = df2['stndrdzed_company_name'].str.replace('[^\w\s]','')

In [19]:
# merge with discovery org data in order to find matches
merged_rows = pd.merge(df, df2, how='inner', on=['stndrdzed_company_name'], suffixes = ['1','2'])

# number of rows after merging

print('# of rows left: %s' % merged_rows.shape[0])

# of rows left: 98811


So, after our intial cleaning process, we had 456,521 rows in our lytics file.  Our DiscoverOrg file had 68,735 rows.  By merging the two files on company name we were able to match 98,811 rows.  That is not a bad start.

Next, we will write the merged and non-merged rows to a file for further analysis.

In [32]:
merged_rows = pd.merge(df, df2, how= 'left', on= 'stndrdzed_company_name', sort=True, suffixes=('_a', '_b'),)

In [33]:
merged_rows

company                                       company_name  \
0          NaN  \t2 Year Experience Soncepts Software Solutio...   
1          NaN       •••  ACTIVELY SEEKING NEW OPPORTUNITIES  •••   
2          NaN                                  @ A Crossroad Ltd   
3          NaN                            - busy bees llcSelect -   
4          NaN                                       -- Choose --   
5          NaN                                              & Co.   
6          NaN                                      (*) GBG, Inc.   
7          NaN                                     ( iii ) DESIGN   
8          NaN           ( J & M Construction: currently defunct)   
9          NaN                              🌎 MIL LOGISTICS, LLC.   
10         NaN                                    ..... On Hiatus   
11         NaN                                    & Other Stories   
12         NaN                                    & Other Stories   
13         NaN                                           ✪ Qmatic   
14         NaN                                   - Select Title -   
15         NaN                                    | Self Employed   
16         NaN                                            - Wiley   
17         NaN                                         -- Year --   
18         NaN                                         -- Year --   
19         NaN  , You Are Loved Foods (Sweetly Changing Lives ...   
20         NaN                             0 to 60 communications   
21         NaN                                         "0" to Now   
22         NaN                      00Diva Digital Communications   
23         NaN                                  01 Jewltd Limited   
24         NaN                                             01game   
25         NaN                                          01jewelry   
26         NaN                                  01MWe Engineering   
27         NaN                                    01NetSmartDeals   
28         NaN                                         03/09/1994   
29         NaN                                         04/06/1989   
...        ...                                                ...   
457437     NaN                                                不识趣   
457438     NaN                                               個人創業   
457439     NaN                                               元智大學   
457440     NaN                                     勞動部勞動力發展署桃竹苗分署   
457441     NaN                                     北京宏泰天汇建筑科技有限公司   
457442     NaN                                                 华为   
457443     NaN                                             华北电力大学   
457444     NaN             喜从天降邀您添加【Q】2810009279链接334948com领取188圆   
457445     NaN                                      四川九州源生物科技有限公司   
457446     NaN                            大脑银行 Manufacturers Inc.   
457447     NaN                                              徐渭工作室   
457448     NaN                                               愛光聖義   
457449     NaN                                       杭州速线网络科技有限公司   
457450     NaN                                                林雨声   
457451     NaN                                             株式会社佐藤   
457452     NaN                                                澳维殴   
457453     NaN                                             華陽優質搬家   
457454     NaN       註冊送《38,380咳出款》地質116458點com，詳情家企鵝3090434764領取   
457455     NaN                                      財團法人食品工業發展研究所   
457456     NaN                                               野炊食堂   
457457     NaN                                               長璽代書   
457458     NaN                                             阿尔斯通运输   
457459     NaN                                       陕西睿融项目管理有限公司   
457460     NaN                 雪国アグリ株式会社/Yukiguni Aguri. Co.,Ltd.   
457461     NaN                                                 駒場   
457462     NaN                                               앵커리어   
45746

In [34]:
import os
path=r'/Users/sasanbahadaran/Downloads'
merged_rows.to_csv(os.path.join(path,r'lytics_profile_disc_org_merged_rows.csv'), index=False)

Next, we can work on wrangling other fields and joining them to the DiscoveryOrg data for enhancement, such as:
* email domain
* address

We could also decide to work on further cleaning up the company data as well.  This could be through performing additional wrangling after examining our output file, or going beyond the deterministic types of methods we have covered so far.  All in all though, the more we standardize our dataset, the better results we will get when performing analysis on our data.